src/Microsoft.AspNetCore.SystemWebAdapters/PrincipalUserFeature.cs

```diff
@@ -0,0 +1,25 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Security.Claims;
+using System.Security.Principal;
+using Microsoft.AspNetCore.Http.Features.Authentication;
+using Microsoft.AspNetCore.SystemWebAdapters;
+
+namespace System.Web;
+
+internal sealed class PrincipalUserFeature : IPrincipalUserFeature, IHttpAuthenticationFeature
+{
+    public IPrincipal? User { get; set; }
+
+    ClaimsPrincipal? IHttpAuthenticationFeature.User
+    {
+        get => User switch
+        {
+            null => null,
+            ClaimsPrincipal claims => claims,
+            IPrincipal user => new ClaimsPrincipal(user),
```

Comment:
Should this be cached?

Reply:
For now, I don't think we need to. It's only used here in the case of a custom IPrincipal, which should be a somewhat fringe case and one we'd like them to move away from. I'll add some logging to flag these cases to le people know, as well as add a section in the docs to explain in more detail.